src/types.ts

```diff
@@ -1,11 +1,43 @@
-export { Request as ExpressRequest } from 'express';
+type Timestamp = number;
 
-export enum Products {
-  jira = 'jira',
-  confluence = 'confluence',
-  bitbucket = 'bitbucket',
+export const ContextQsh = 'context-qsh';
+
+/**
+ * Represents an extensible payload of a Connect JWT.
+ */
+export interface ConnectJwt {
+  sub: string;
+  qsh?: string | typeof ContextQsh;
+  iss: string; // clientKey
+  aud: string; // app's baseUrl
+  exp: Timestamp;
+  iat: Timestamp;
+  context: unknown;
+  [key: string]: unknown;
+}
+
+export enum InstallationType {
+  newInstallation = 'newInstallation',
+  update = 'update',
 }
 
-export type TokenPayload = Record<string, unknown>;
+/**
+ * Holds current credentials and the entity representing an existing installation.
+ */
+export interface CredentialsWithEntity<E> {
+  sharedSecret: string;
+  storedEntity: E;
+}
+
+/**
+ * Used to load data from an existing installation.
+ */
+export interface CredentialsLoader<E> {
+  (clientKey: string): Promise<CredentialsWithEntity<E> | undefined>;
+}
 
-export type CustomTokenExtractor = () => string | undefined;
+/**
+ * Chooses the algorithm used to verify the content of an incoming request from Connect.
+ */
+export type QueryStringHashType = 'computed' | 'context' | 'any' | 'skip';
+export type InstallationQueryStringHashType = 'computed' | 'skip';
```

Comment:
In which cases the consumer can set `any` or `skip`? IIRC with a recent fix for QSH vulnerability, there's no case the QSH validation can be skipped. It's always `computed` or `context` if the token comes from the front-end via `window.AP.context.getToken()`.

Reply:
Perhaps for testing. Not 100% TBH. I just wanted to be on par with ACE: https://bitbucket.org/atlassian/atlassian-connect-express/src/d3e4ba420a495900c7f15cf4d8d74f189f2e2f62/lib/middleware/authentication.js#lines-84